mlflow/langchain/utils/chat.py

```diff
@@ -277,22 +280,33 @@ def json_dict_might_be_chat_request(json_message: dict):
             # Additional keys can't be specified when calling LangChain invoke() / batch()
             # with chat messages
             len(json_message) == 1
+            # messages field should be a list
+            and isinstance(json_message["messages"], list)
         )
 
-    if isinstance(request_json, dict) and json_dict_might_be_chat_request(request_json):
+    if not should_transform_requst_json_for_chat(lc_model):
+        return request_json, False
+
+    if json_dict_might_be_chat_request(request_json):
         try:
-            return _convert_chat_request_or_throw(request_json), True
-        except pydantic.ValidationError:
+            result = _convert_chat_request_or_throw(request_json)
+            if hasattr(lc_model, "input_schema"):
+                # sometimes the input schema is not reliable and we should try
+                # to invoke the model with the request instead, restricting the check to only
+                # string input schemas to reduce the blast radius
+                # example: test_save_load_chain_as_code
+                if lc_model.input_schema.schema().get("type", "string") == "string":
+                    # TODO: migrate this logic inside APIRequest to avoid invoke twice
+                    lc_model.invoke(result)
+                    return result, True
+                else:
+                    lc_model.input_schema.validate(result)
+                    return result, True
             return request_json, False
-    elif isinstance(request_json, list) and all(
-        json_dict_might_be_chat_request(json) for json in request_json
-    ):
-        try:
-            return (
-                [_convert_chat_request_or_throw(json_dict) for json_dict in request_json],
-                True,
-            )
-        except pydantic.ValidationError:
+        # we should catch all exceptions here including pydantic validation errors
```

Comment:
With this removal, are we simply removing a completely broken condition or will this have any breaking changes?

Reply:
request_json inside APIRequest should always be dictionaries, previously we have list[dict] here because it's the batch input data.
e.g. {"messages": ...} continues to be a valid input;
[{"messages": ...}, {"messages": ...}] will be tackled in the threadpool, and each request is still a dictionary;
[[{"messages": ...}]] will not be a valid input, because it expects the model to work on [{"messages": ...}] which is not true anyways.